src/inmanta/data/__init__.py

```diff
@@ -4203,6 +4277,116 @@ async def query_resource_actions(
             async with con.transaction():
                 return [cls(**record, from_postgres=True) async for record in con.cursor(query, *values)]
 
+    @classmethod
+    async def get_resource_events(
+        cls, env: Environment, resource_id: "resources.Id", exclude_change: Optional[const.Change] = None
+    ) -> Dict[ResourceIdStr, List["ResourceAction"]]:
+        """
+        Get all events that should be processed by this specific resource, for the current deployment
+
+        This method searches across versions!
+
+        This means:
+        1. assure a deployment is ongoing
+        2. get the time range between the start of this deployment and the last successfull deploy
+        3. get all resources required by this resource
+        4. get all resource actions of type deploy emitted by the resource of step 3 in the time interval of step 2
+
+        :param env: environment to consider
+        :param resource_id: resource to consider, should be in deploying state
+        :param exclude_change: in step 4, exclude all resource actions with this specific type of change
+        """
+
+        # This is bang on the critical path for the agent
+        # Squeeze out as much performance from postgresql as we can
+
+        # steps 1 and 2:
+        # find the interval between the current deploy and the previous successful deploy
+        # also check we are currently deploying
+        # do all of this in one query
+        resource_id_str = resource_id.resource_version_str()
+
+        # These two variables are actually of type datetime.datetime
+        # but mypy doesn't know as they come from the DB
+        # mypy also doesn't care, because they go back into the DB
+        current_deploy_start: object
+        last_deploy_start: Optional[object]
+
+        end_query = """
+        with
+            base_ra as (
+            SELECT ra.*
+                FROM public.resourceaction_resource as jt
+                    INNER JOIN public.resourceaction as ra
+                        ON ra.action_id = jt.resource_action_id
+                    WHERE jt.environment=$1 AND ra.environment=$1 AND jt.resource_id=$2::varchar AND ra.action='deploy'
+                    ORDER BY ra.started DESC
+            )
+        SELECT
+            (SELECT started from base_ra ORDER BY started DESC LIMIT 1) as begin_started,
+            (SELECT status from base_ra ORDER BY started DESC LIMIT 1) as begin_status,
+            COALESCE((SELECT started from base_ra where status='deployed' ORDER BY started DESC LIMIT 1), NULL) as started
+        """
+
+        async with cls.get_connection() as connection:
+            result = await connection.fetchrow(end_query, env.id, resource_id.resource_str())
+
+            if not result or result["begin_status"] is None:
+                raise BadRequest(
```

Comment:
I don't think we should raise `inmanta.protocol` exceptions here.

Reply:
I agree, but failed to come up with a better solution or rationale why this would be bad.